### DISCUSSION TITLE
Request Matching with '='

### DISCUSSION THREAD

**Discussion by ashanka-singh-qatalog** [fauxauldrich/camouflage#64] [Q&A]
Hi Shubhendu, 
Greetings! 

I have recently started using https://testinggospels.github.io/camouflage/  and I am finding it very useful. 
I have one query here, In the request matching section [Request-match](https://github.com/testinggospels/camouflage/blob/develop/docs/request-matching.md#explanation)

When I try to perform` {{#if request.query.task='1234'}}` in my `GET.mock` file I see below error:

```Expecting 'CLOSE_RAW_BLOCK', 'CLOSE', 'CLOSE_UNESCAPED', 'OPEN_SEXPR', 'CLOSE_SEXPR', 'ID', 'OPEN_BLOCK_PARAMS', 'STRING', 'NUMBER', 'BOOLEAN', 'UNDEFINED', 'NULL', 'DATA', 'SEP', got 'EQUALS'.```

Do you have any suggestions around this? Basically, I have different task-id value and I want to send mock based on the Id. However  `=` doesn't seem to work for me.

**Comment by shubhendumadhukar** [comment]
Hi Ashanka,

Until recently I too was under assumption that this would work, but it appears handlebars' if helper does not support comparisions. (I need to update Camouflage docs)

See https://stackoverflow.com/questions/34252817/handlebarsjs-check-if-a-string-is-equal-to-a-value

There are two ways to achieve your use case:
- Use either `code` or `inject` helper. See example in this [issue](https://github.com/testinggospels/camouflage/issues/59); OR
- Create your own custom helper and load it while starting Camouflage

Since this is such a common use case, I think it would be best to have a helper included as part of Camouflage. It would be great if you could create a feature request issue for the same.

**Reply by shubhendumadhukar** [reply]
I worked out a quick and dirty way of loading an external custom helper for this use case.

- If you already have a JSON file in your project to load external handlerbars, add following json object to it. If not, set it up referring to this [guide](https://testinggospels.github.io/camouflage/external-helper.html)
```json
  {
    "name": "is",
    "logic":"(()=>{ logger.info(JSON.stringify((context.hash))); if(context.hash.value1===context.hash.value2) {return context.fn(this);} else {return context.inverse(this);} })()"
  }
```
- Restart Camouflage
- Use following as content of your mock file
```
HTTP/1.1 200 OK

{{#is value1=request.query.name value2='Shubhendu'}}
   Response if true
{{else}}
   Response if false
{{/is}}
```
- Modify value1 and value2 accordingly.

Let me know if this works!

**Reply by ashanka-singh-qatalog** [reply]
Hi Shubhendu! Thanks! That worked perfectly.

Just something I noticed. If I try to evaluate this for a POST request with the the value coming from `request.body.name` instead of `request.query.name`, the check works the first time, but all subsequent calls end up with the same result as the first run irrespective of the `request.body.name` value from subsequent calls.

For instance,

custom_handlerbars.json:
```json
[
  {
    "name": "is",
    "logic": "(()=>{ logger.info(JSON.stringify((context.hash))); if(request.body.name===context.hash.value2) {return context.fn(this);} else {return context.inverse(this);} })()"
  }
]
```

POST.mock:
```
HTTP/1.1 200 OK
Content-Type: application/json; charset=UTF-8

{{#is value2='Shubhendu'}}
   {"branch": "if"}
{{else}}
   {"branch": "else"}
{{/is}}
```

```bash
$ echo '{"name": "Shubhendu"}' | curl -XPOST -H "Content-Type: application/json" http://localhost:8080/foo
{"branch": "if"}

$ echo '{"name": "Something"}' | curl -XPOST -H "Content-Type: application/json" http://localhost:8090/foo
{"branch": "if"}
```

Now, if I restart camouflage and try the same tests in the reverse order:
```bash
$ echo '{"name": "Something"}' | curl -XPOST -H "Content-Type: application/json" http://localhost:8090/foo
{"branch": "else"}

$ echo '{"name": "Shubhendu"}' | curl -XPOST -H "Content-Type: application/json" http://localhost:8090/foo
{"branch": "else"}
```

Just something I noticed. Your solution works perfectly for my requirements. Thank you!

**Reply by shubhendumadhukar** [reply]
I'd recommend upgrading to 0.5.0. It was an issue ([#59 for context](https://github.com/testinggospels/camouflage/issues/59)) due to cache. In 0.5.0, you can disable cache in config.yml. That should help.

Also the `is` helper should have following syntax:
```
{{#is value1=request.body.name value2='Shubhendu'}}
```
I assuming this was a typo in your previous comment.

**Comment by shubhendumadhukar** [comment]
[`is` helper](https://testinggospels.github.io/camouflage/handlebars/#is) is now going to be included with Camouflage's installation, 0.5.11 onwards.

**Reply by ashanka-singh-qatalog** [reply]
> [`is` helper](https://testinggospels.github.io/camouflage/handlebars/#is) is now going to be included with Camouflage's installation, 0.5.11 onwards.

This is **Awsome** !! thanks a lot for doing this.
